flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeSubpartitionReader.java

```diff
@@ -176,21 +202,28 @@ public void resumeConsumption() {
 
     @Override
     public Throwable getFailureCause() {
-        // we can never throw an error after this was created
-        return null;
+        synchronized (lock) {
+            return failureCause;
+        }
     }
 
     @Override
     public boolean isAvailable(int numCreditsAvailable) {
-        if (numCreditsAvailable > 0) {
-            return !buffersRead.isEmpty();
-        }
+        synchronized (lock) {
+            if (isReleased) {
+                return true;
+            }
 
-        return !buffersRead.isEmpty() && !buffersRead.peek().isBuffer();
+            if (numCreditsAvailable > 0) {
+                return !buffersRead.isEmpty();
+            }
+
+            return !buffersRead.isEmpty() && !buffersRead.peek().isBuffer();
+        }
     }
 
     @Override
     public int unsynchronizedGetNumberOfQueuedBuffers() {
```

Comment:
It's out of the scope of this change. But I think we need some java docs for the methods of `ResultSubpartitionView`. At the moment I can hardly understand the definitions of some of them without looking into their invocations.

Reply:
This method is only for metric. I agree we should add some java doc to methods of ResultSubpartitionView. But that is out of the scope of this PR. Maybe we can do the improvement in another PR in the future.